include/api/CDataFrameAnalysisInstrumentation.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_api_CDataFrameAnalysisInstrumentation_h
+#define INCLUDED_ml_api_CDataFrameAnalysisInstrumentation_h
+
+#include <core/CProgramCounters.h>
+#include <core/CRapidJsonConcurrentLineWriter.h>
+
+#include <maths/CDataFrameAnalysisInstrumentationInterface.h>
+
+#include <api/CDataFrameAnalysisInstrumentation.h>
+#include <api/ImportExport.h>
+
+#include <atomic>
+#include <cstdint>
+
+namespace ml {
+namespace api {
+
+//! \brief Instrumentation class for collecting data frame analysis job statistics.
+//!
+//! DESCRIPTION:\n
+//! Responsible for collecting data frame analysis job statistics, i.e. memory usage,
+//! progress, parameters, quality of results. The class also implements the functionality to
+//! write the state at different iteration into the results pipe.
+//!
+//! The implementation is based on a queue that can contain max. 10 element. This is needed since  the
+//! analysis starts asynchronously before the result writer (consumer) is created. The time difference
+//! between the two events is negligible, nevertheless, if the queue is full, new elements will be dropped.
```

Comment:
This comment should probably change if we no longer have a queue.

Reply:
Done